Connecting to carddav radicale server

Hello, I am aware that the library is for webdav but just to be sure, is it possible to use it to connect to carddav server?

Thank you.

Thank you for the kind reply.

After trying it out for a bit using this snippet code

```
            Sardine sardine = new OkHttpSardine();
            sardine.setCredentials(username, password);

            try {
                List<DavResource> resources = sardine.list(url);
                Log.d("testcarddav", resources.toString());
            } catch (IOException e) {
                e.printStackTrace();
            }
```

I've encountered this issue:

Challenges [realm:  radicale - "password required" charset="java.nio.charset.charseticu[iso-8859-1]"]

```
   at com.thegrizzlylabs.sardineandroid.util.SardineUtil.unmarshal(SardineUtil.java:140)
        at com.thegrizzlylabs.sardineandroid.impl.handler.MultiStatusResponseHandler.getMultistatus(MultiStatusResponseHandler.java:38)
        at com.thegrizzlylabs.sardineandroid.impl.handler.MultiStatusResponseHandler.handleResponse(MultiStatusResponseHandler.java:27)
        at com.thegrizzlylabs.sardineandroid.impl.handler.ResourcesResponseHandler.handleResponse(ResourcesResponseHandler.java:25)
        at com.thegrizzlylabs.sardineandroid.impl.handler.ResourcesResponseHandler.handleResponse(ResourcesResponseHandler.java:19)
        at com.thegrizzlylabs.sardineandroid.impl.OkHttpSardine.execute(OkHttpSardine.java:595)
        at com.thegrizzlylabs.sardineandroid.impl.OkHttpSardine.propfind(OkHttpSardine.java:180)
        at com.thegrizzlylabs.sardineandroid.impl.OkHttpSardine.list(OkHttpSardine.java:149)
        at com.thegrizzlylabs.sardineandroid.impl.OkHttpSardine.list(OkHttpSardine.java:125)
        at com.thegrizzlylabs.sardineandroid.impl.OkHttpSardine.list(OkHttpSardine.java:120)
        at com.example.pierson.testcarddav.MainActivity$TestCardDav.doInBackground(MainActivity.java:37)
        at com.example.pierson.testcarddav.MainActivity$TestCardDav.doInBackground(MainActivity.java:28)
        at android.os.AsyncTask$2.call(AsyncTask.java:288)
        at java.util.concurrent.FutureTask.run(FutureTask.java:237)
        at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:231)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1112)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:587)
        at java.lang.Thread.run(Thread.java:818)
    Caused by: org.simpleframework.xml.core.ElementException: Element 'href' does not have a match in class com.thegrizzlylabs.sardineandroid.model.PrincipalCollectionSet at line 2
09-03 17:55:43.859 20914-20937/com.example.pierson.testcarddav W/System.err:     at org.simpleframework.xml.core.Composite.readElement(Composite.java:527)
        at org.simpleframework.xml.core.Composite.readElements(Composite.java:445)
        at org.simpleframework.xml.core.Composite.access$400(Composite.java:59)
        at org.simpleframework.xml.core.Composite$Builder.read(Composite.java:1383)
        at org.simpleframework.xml.core.Composite.read(Composite.java:201)
        at org.simpleframework.xml.core.Composite.read(Composite.java:148)
        at org.simpleframework.xml.core.Traverser.read(Traverser.java:92)
09-03 17:55:43.864 20914-20937/com.example.pierson.testcarddav W/System.err:     at org.simpleframework.xml.core.Persister.read(Persister.java:625)
        at org.simpleframework.xml.core.Persister.read(Persister.java:606)
        at org.simpleframework.xml.core.Persister.read(Persister.java:584)
        at org.simpleframework.xml.core.Persister.read(Persister.java:480)
        at com.thegrizzlylabs.sardineandroid.model.Prop$PropConverter.read(Prop.java:335)
        at com.thegrizzlylabs.sardineandroid.model.Prop$PropConverter.read(Prop.java:306)
        at org.simpleframework.xml.convert.RegistryStrategy.read(RegistryStrategy.java:129)
        at org.simpleframework.xml.convert.RegistryStrategy.read(RegistryStrategy.java:108)
        at org.simpleframework.xml.core.Source.getOverride(Source.java:360)
        at org.simpleframework.xml.core.Factory.getConversion(Factory.java:207)
        at org.simpleframework.xml.core.Factory.getOverride(Factory.java:139)
        at org.simpleframework.xml.core.ObjectFactory.getInstance(ObjectFactory.java:63)
        at org.simpleframework.xml.core.Composite.read(Composite.java:139)
        at org.simpleframework.xml.core.Composite.readVariable(Composite.java:623)
        at org.simpleframework.xml.core.Composite.readInstance(Composite.java:573)
        at org.simpleframework.xml.core.Composite.readUnion(Composite.java:549)
        at org.simpleframework.xml.core.Composite.readElement(Composite.java:532)
09-03 17:55:43.869 20914-20937/com.example.pierson.testcarddav W/System.err:     at org.simpleframework.xml.core.Composite.readElements(Composite.java:445)
        at org.simpleframework.xml.core.Composite.access$400(Composite.java:59)
        at org.simpleframework.xml.core.Composite$Builder.read(Composite.java:1383)
        at org.simpleframework.xml.core.Composite.read(Composite.java:201)
        at org.simpleframework.xml.core.Composite.read(Composite.java:148)
        at org.simpleframework.xml.core.Traverser.read(Traverser.java:92)
        at org.simpleframework.xml.core.CompositeInlineList.read(CompositeInlineList.java:190)
        at org.simpleframework.xml.core.CompositeInlineList.read(CompositeInlineList.java:167)
        at org.simpleframework.xml.core.CompositeInlineList.read(CompositeInlineList.java:124)
        at org.simpleframework.xml.core.Composite.readVariable(Composite.java:623)
        at org.simpleframework.xml.core.Composite.readInstance(Composite.java:573)
09-03 17:55:43.874 20914-20937/com.example.pierson.testcarddav W/System.err:     at org.simpleframework.xml.core.Composite.readUnion(Composite.java:549)
        at org.simpleframework.xml.core.Composite.readElement(Composite.java:532)
        at org.simpleframework.xml.core.Composite.readElements(Composite.java:445)
        at org.simpleframework.xml.core.Composite.access$400(Composite.java:59)
        at org.simpleframework.xml.core.Composite$Builder.read(Composite.java:1383)
        at org.simpleframework.xml.core.Composite.read(Composite.java:201)
        at org.simpleframework.xml.core.Composite.read(Composite.java:148)
        at org.simpleframework.xml.core.Traverser.read(Traverser.java:92)
09-03 17:55:43.874 20914-20937/com.example.pierson.testcarddav W/System.err:     at org.simpleframework.xml.core.CompositeInlineList.read(CompositeInlineList.java:190)
        at org.simpleframework.xml.core.CompositeInlineList.read(CompositeInlineList.java:167)
        at org.simpleframework.xml.core.CompositeInlineList.read(CompositeInlineList.java:124)
        at org.simpleframework.xml.core.Composite.readVariable(Composite.java:623)
        at org.simpleframework.xml.core.Composite.readInstance(Composite.java:573)
09-03 17:55:43.879 20914-20937/com.example.pierson.testcarddav W/System.err:     at org.simpleframework.xml.core.Composite.readUnion(Composite.java:549)
        at org.simpleframework.xml.core.Composite.readElement(Composite.java:532)
        at org.simpleframework.xml.core.Composite.readElements(Composite.java:445)
        at org.simpleframework.xml.core.Composite.access$400(Composite.java:59)
        at org.simpleframework.xml.core.Composite$Builder.read(Composite.java:1383)
        at org.simpleframework.xml.core.Composite.read(Composite.java:201)
        at org.simpleframework.xml.core.Composite.read(Composite.java:148)
        at org.simpleframework.xml.core.Traverser.read(Traverser.java:92)
        at org.simpleframework.xml.core.Persister.read(Persister.java:625)
        at org.simpleframework.xml.core.Persister.read(Persister.java:606)
        at org.simpleframework.xml.core.Persister.read(Persister.java:584)
        at org.simpleframework.xml.core.Persister.read(Persister.java:543)
        at org.simpleframework.xml.core.Persister.read(Persister.java:444)
        at com.thegrizzlylabs.sardineandroid.util.SardineUtil.unmarshal(SardineUtil.java:135)
    	... 17 more
```

I'm guessing it's something to do with the password?

--- a/src/main/java/com/thegrizzlylabs/sardineandroid/impl/OkHttpSardine.java
+++ b/src/main/java/com/thegrizzlylabs/sardineandroid/impl/OkHttpSardine.java
@@ -563,7 +563,7 @@ public class OkHttpSardine implements Sardine {
                         if (propstat.getProp() != null
                                 && propstat.getProp().getPrincipalCollectionSet() != null
                                 && propstat.getProp().getPrincipalCollectionSet().getHref() != null) {
-                            collections.addAll(propstat.getProp().getPrincipalCollectionSet().getHref());
+                            collections.add(propstat.getProp().getPrincipalCollectionSet().getHref());
                         }
                     }
                 }

--- a/src/main/java/com/thegrizzlylabs/sardineandroid/model/CurrentUserPrivilegeSet.java
+++ b/src/main/java/com/thegrizzlylabs/sardineandroid/model/CurrentUserPrivilegeSet.java
@@ -1,0 +1,13 @@
+package com.thegrizzlylabs.sardineandroid.model;
+
+import org.simpleframework.xml.ElementList;
+import org.simpleframework.xml.Root;
+
+import java.util.List;
+
+@Root
+public class CurrentUserPrivilegeSet {
+
+    @ElementList(inline = true)
+    public List<Privilege> privileges;
+}

--- a/src/main/java/com/thegrizzlylabs/sardineandroid/model/EntityWithAnyElement.java
+++ b/src/main/java/com/thegrizzlylabs/sardineandroid/model/EntityWithAnyElement.java
@@ -1,0 +1,10 @@
+package com.thegrizzlylabs.sardineandroid.model;
+
+import org.w3c.dom.Element;
+
+import java.util.List;
+
+public interface EntityWithAnyElement {
+
+    public List<Element> getAny();
+}

--- a/src/main/java/com/thegrizzlylabs/sardineandroid/model/PrincipalCollectionSet.java
+++ b/src/main/java/com/thegrizzlylabs/sardineandroid/model/PrincipalCollectionSet.java
@@ -8,10 +8,9 @@
 
 package com.thegrizzlylabs.sardineandroid.model;
 
+import org.simpleframework.xml.Element;
 import org.simpleframework.xml.Namespace;
 import org.simpleframework.xml.Root;
-
-import java.util.List;
 
 
 /**
@@ -37,15 +36,14 @@ import java.util.List;
 @Namespace(prefix = "D", reference = "DAV:")
 public class PrincipalCollectionSet {
 
-	private List<String> href;
+	@Element(required = false)
+	protected String href;
 
- 
-
-	public List<String> getHref() {
+	public String getHref() {
 		return href;
 	}
 
-	public void setHref(List<String> href) {
+	public void setHref(String href) {
 		this.href = href;
 	}
 

--- a/src/main/java/com/thegrizzlylabs/sardineandroid/model/PrincipalURL.java
+++ b/src/main/java/com/thegrizzlylabs/sardineandroid/model/PrincipalURL.java
@@ -8,6 +8,7 @@
 
 package com.thegrizzlylabs.sardineandroid.model;
 
+import org.simpleframework.xml.Element;
 import org.simpleframework.xml.Namespace;
 import org.simpleframework.xml.Root;
 
@@ -34,6 +35,7 @@ import org.simpleframework.xml.Root;
 @Namespace(prefix = "D", reference = "DAV:")
 public class PrincipalURL {
 
+	@Element
 	private String href;
 
  

--- a/src/main/java/com/thegrizzlylabs/sardineandroid/model/Prop.java
+++ b/src/main/java/com/thegrizzlylabs/sardineandroid/model/Prop.java
@@ -9,22 +9,12 @@
 package com.thegrizzlylabs.sardineandroid.model;
 
 
-import com.thegrizzlylabs.sardineandroid.ElementConverter;
-import com.thegrizzlylabs.sardineandroid.util.SardineUtil;
-
 import org.simpleframework.xml.Element;
 import org.simpleframework.xml.Namespace;
 import org.simpleframework.xml.Root;
-import org.simpleframework.xml.Serializer;
-import org.simpleframework.xml.convert.Converter;
-import org.simpleframework.xml.stream.InputNode;
-import org.simpleframework.xml.stream.OutputNode;
 
-import java.lang.reflect.Field;
 import java.util.ArrayList;
-import java.util.HashMap;
 import java.util.List;
-import java.util.Map;
 
 
 /**
@@ -59,7 +49,7 @@ import java.util.Map;
 
 @Root(strict = false)
 @Namespace(prefix = "D", reference = "DAV:")
-public class Prop {
+public class Prop implements EntityWithAnyElement {
 
     @Element(required = false)
     protected String creationdate;
@@ -101,6 +91,10 @@ public class Prop {
 
     public Resourcetype getResourcetype() {
         return resourcetype;
+    }
+
+    public void setResourcetype(Resourcetype resourcetype) {
+        this.resourcetype = resourcetype;
     }
 
     public String getCreationdate() {
@@ -159,6 +153,14 @@ public class Prop {
         this.getlastmodified = getlastmodified;
     }
 
+    public void setSupportedlock(Supportedlock supportedlock) {
+        this.supportedlock = supportedlock;
+    }
+
+    public Supportedlock getSupportedlock() {
+        return supportedlock;
+    }
+
     //ACL elements
     @Element(required = false)
     protected Owner owner;
@@ -171,10 +173,13 @@ public class Prop {
 
 
     @Element(name="principal-collection-set", required = false)
-    protected PrincipalCollectionSet principalCollectionSet;
+    private PrincipalCollectionSet principalCollectionSet;
 
-    @Element(name="principal-URL", required = false)
-    protected PrincipalURL principalURL;
+    @Element(name="current-user-principal", required = false)
+    private PrincipalURL principalURL;
+
+    @Element(name="current-user-privilege-set")
+    private CurrentUserPrivilegeSet currentUserPrivilegeSet;
 
     /**
      * Gets the value of the lockdiscovery property.
@@ -256,6 +261,7 @@ public class Prop {
      * Objects of the following type(s) are allowed in the list
      * {@link Element }
      */
+    @Override
     public List<org.w3c.dom.Element> getAny() {
         if (any == null) {
             any = new ArrayList<>();
@@ -303,64 +309,11 @@ public class Prop {
 		this.principalURL = principalURL;
 	}
 
-    public static class PropConverter implements Converter<Prop> {
+    public void setCurrentUserPrivilegeSet(CurrentUserPrivilegeSet currentUserPrivilegeSet) {
+        this.currentUserPrivilegeSet = currentUserPrivilegeSet;
+    }
 
-        private Serializer serializer;
-
-        public PropConverter(Serializer serializer) {
-            this.serializer = serializer;
-        }
-
-        private Map<String, Field> getPropElements() {
-            Map<String, Field> elementsFields = new HashMap<>();
-            for (Field field : Prop.class.getDeclaredFields()) {
-                Element fieldAnnotation = field.getAnnotation(Element.class);
-                if (fieldAnnotation != null) {
-                    String name = fieldAnnotation.name().equals("") ? field.getName() : fieldAnnotation.name();
-                    elementsFields.put(name, field);
-                }
-            }
-            return elementsFields;
-        }
-
-        @Override
-        public Prop read(InputNode node) throws Exception {
-            Map<String, Field> propElements = getPropElements();
-            Prop prop = new Prop();
-            List<org.w3c.dom.Element> anyElements = prop.getAny();
-            InputNode childNode;
-            while((childNode = node.getNext()) != null) {
-                if (propElements.containsKey(childNode.getName())) {
-                    Field field = propElements.get(childNode.getName());
-                    field.set(prop, serializer.read(field.getType(), childNode));
-                } else {
-                    org.w3c.dom.Element element = ElementConverter.read(childNode);
-                    anyElements.add(element);
-                }
-            }
-            return prop;
-        }
-
-        @Override
-        public void write(OutputNode node, Prop prop) throws Exception {
-            for(org.w3c.dom.Element domElement : prop.getAny()) {
-                ElementConverter.write(node, domElement);
-            }
-            Map<String, Field> propElements = getPropElements();
-            for (String fieldName : propElements.keySet()) {
-                Field field = propElements.get(fieldName);
-                Object value = field.get(prop);
-                if (value == null) {
-                    continue;
-                }
-                if (value instanceof String) {
-                    OutputNode childNode = node.getChild(fieldName);
-                    childNode.setReference(SardineUtil.DEFAULT_NAMESPACE_URI);
-                    childNode.setValue((String)value);
-                } else {
-                    serializer.write(value, node);
-                }
-            }
-        }
+    public CurrentUserPrivilegeSet getCurrentUserPrivilegeSet() {
+        return currentUserPrivilegeSet;
     }
 }

--- a/src/main/java/com/thegrizzlylabs/sardineandroid/model/Property.java
+++ b/src/main/java/com/thegrizzlylabs/sardineandroid/model/Property.java
@@ -8,7 +8,7 @@
 
 package com.thegrizzlylabs.sardineandroid.model;
 
-import com.thegrizzlylabs.sardineandroid.ElementConverter;
+import com.thegrizzlylabs.sardineandroid.util.ElementConverter;
 
 import org.simpleframework.xml.Namespace;
 import org.simpleframework.xml.Root;

--- a/src/main/java/com/thegrizzlylabs/sardineandroid/model/Resourcetype.java
+++ b/src/main/java/com/thegrizzlylabs/sardineandroid/model/Resourcetype.java
@@ -8,9 +8,9 @@
 
 package com.thegrizzlylabs.sardineandroid.model;
 
+import org.simpleframework.xml.Element;
 import org.simpleframework.xml.Namespace;
 import org.simpleframework.xml.Root;
-import org.w3c.dom.Element;
 
 import java.util.ArrayList;
 import java.util.List;
@@ -38,14 +38,14 @@ import java.util.List;
  */
 @Root
 @Namespace(prefix = "D", reference = "DAV:")
-public class Resourcetype {
+public class Resourcetype implements EntityWithAnyElement {
 
-    @org.simpleframework.xml.Element(required = false)
+    @Element(required = false)
     private Collection collection;
 
     private Principal principal;
-    //@XmlAnyElement
-    private List<Element> any;
+
+    private List<org.w3c.dom.Element> any;
 
     /**
      * Gets the value of the collection property.
@@ -101,9 +101,10 @@ public class Resourcetype {
      * 
      * 
      */
-    public List<Element> getAny() {
+	@Override
+    public List<org.w3c.dom.Element> getAny() {
         if (any == null) {
-            any = new ArrayList<Element>();
+            any = new ArrayList<>();
         }
         return this.any;
     }

--- a/src/main/java/com/thegrizzlylabs/sardineandroid/util/ElementConverter.java
+++ b/src/main/java/com/thegrizzlylabs/sardineandroid/util/ElementConverter.java
@@ -1,6 +1,4 @@
-package com.thegrizzlylabs.sardineandroid;
-
-import com.thegrizzlylabs.sardineandroid.util.SardineUtil;
+package com.thegrizzlylabs.sardineandroid.util;
 
 import org.simpleframework.xml.stream.InputNode;
 import org.simpleframework.xml.stream.OutputNode;

--- a/src/main/java/com/thegrizzlylabs/sardineandroid/util/EntityWithAnyElementConverter.java
+++ b/src/main/java/com/thegrizzlylabs/sardineandroid/util/EntityWithAnyElementConverter.java
@@ -1,0 +1,99 @@
+package com.thegrizzlylabs.sardineandroid.util;
+
+import com.thegrizzlylabs.sardineandroid.model.EntityWithAnyElement;
+
+import org.simpleframework.xml.Element;
+import org.simpleframework.xml.Serializer;
+import org.simpleframework.xml.convert.Converter;
+import org.simpleframework.xml.stream.InputNode;
+import org.simpleframework.xml.stream.OutputNode;
+
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class EntityWithAnyElementConverter<T extends EntityWithAnyElement> implements Converter<T> {
+
+    private Class<T> entityClass;
+    private Serializer serializer;
+
+    public EntityWithAnyElementConverter(Serializer serializer, Class<T> entityClass) {
+        this.serializer = serializer;
+        this.entityClass = entityClass;
+    }
+
+    private Map<String, Field> getEntityFields() {
+        Map<String, Field> elementsFields = new HashMap<>();
+        for (Field field : entityClass.getDeclaredFields()) {
+            Element fieldAnnotation = field.getAnnotation(Element.class);
+            if (fieldAnnotation != null) {
+                String name = fieldAnnotation.name().equals("") ? field.getName() : fieldAnnotation.name();
+                elementsFields.put(name, field);
+            }
+        }
+        return elementsFields;
+    }
+
+    private Method getSetterForField(Field field) throws NoSuchMethodException {
+        String fieldName = field.getName();
+        String capitalizedFieldName = fieldName.substring(0,1).toUpperCase() + fieldName.substring(1);
+        return entityClass.getMethod("set" + capitalizedFieldName, field.getType());
+    }
+
+    private Method getGetterForField(Field field) throws NoSuchMethodException {
+        String fieldName = field.getName();
+        String capitalizedFieldName = fieldName.substring(0,1).toUpperCase() + fieldName.substring(1);
+        return entityClass.getMethod("get" + capitalizedFieldName);
+    }
+
+    @Override
+    public T read(InputNode node) throws Exception {
+        Map<String, Field> entityFields = getEntityFields();
+        T entity = entityClass.newInstance();
+        List<org.w3c.dom.Element> anyElements = entity.getAny();
+        InputNode childNode;
+        while((childNode = node.getNext()) != null) {
+            if (entityFields.containsKey(childNode.getName())) {
+                Field field = entityFields.get(childNode.getName());
+                getSetterForField(field).invoke(entity, serializer.read(field.getType(), childNode));
+            } else if (childNode.getPrefix() != null && !childNode.getPrefix().isEmpty()) {
+                org.w3c.dom.Element element = ElementConverter.read(childNode);
+                anyElements.add(element);
+            } else {
+                // Probably a WebDAV field we don't support yet
+                skipChildrenOfNode(childNode);
+            }
+        }
+        return entity;
+    }
+
+    private void skipChildrenOfNode(InputNode node) throws Exception {
+        while(node.getNext() != null) {
+            // Do nothing
+        }
+    }
+
+    @Override
+    public void write(OutputNode node, T entity) throws Exception {
+        for(org.w3c.dom.Element domElement : entity.getAny()) {
+            ElementConverter.write(node, domElement);
+        }
+        Map<String, Field> entityFields = getEntityFields();
+        for (String fieldName : entityFields.keySet()) {
+            Field field = entityFields.get(fieldName);
+            Object value = getGetterForField(field).invoke(entity);
+            if (value == null) {
+                continue;
+            }
+            if (value instanceof String) {
+                OutputNode childNode = node.getChild(fieldName);
+                childNode.setReference(SardineUtil.DEFAULT_NAMESPACE_URI);
+                childNode.setValue((String)value);
+            } else {
+                serializer.write(value, node);
+            }
+        }
+    }
+}

--- a/src/main/java/com/thegrizzlylabs/sardineandroid/util/SardineUtil.java
+++ b/src/main/java/com/thegrizzlylabs/sardineandroid/util/SardineUtil.java
@@ -2,9 +2,9 @@ package com.thegrizzlylabs.sardineandroid.util;
 
 import com.thegrizzlylabs.sardineandroid.model.Prop;
 import com.thegrizzlylabs.sardineandroid.model.Property;
+import com.thegrizzlylabs.sardineandroid.model.Resourcetype;
 
 import org.simpleframework.xml.Serializer;
-import org.simpleframework.xml.convert.Converter;
 import org.simpleframework.xml.convert.Registry;
 import org.simpleframework.xml.convert.RegistryStrategy;
 import org.simpleframework.xml.core.Persister;
@@ -121,9 +121,9 @@ public final class SardineUtil {
         Registry registry = new Registry();
         Strategy strategy = new RegistryStrategy(registry);
         Serializer serializer = new Persister(strategy, format);
-        Converter<Prop> converter = new Prop.PropConverter(serializer);
 
-        registry.bind(Prop.class, converter);
+        registry.bind(Prop.class, new EntityWithAnyElementConverter<>(serializer, Prop.class));
+        registry.bind(Resourcetype.class, new EntityWithAnyElementConverter<>(serializer, Resourcetype.class));
         registry.bind(Property.class, Property.PropertyConverter.class);
 
         return serializer;

--- a/src/test/java/com/thegrizzlylabs/sardineandroid/impl/handler/MultiStatusResponseHandlerTest.java
+++ b/src/test/java/com/thegrizzlylabs/sardineandroid/impl/handler/MultiStatusResponseHandlerTest.java
@@ -20,9 +20,9 @@ import com.thegrizzlylabs.sardineandroid.DavAcl;
 import com.thegrizzlylabs.sardineandroid.DavPrincipal;
 import com.thegrizzlylabs.sardineandroid.DavResource;
 import com.thegrizzlylabs.sardineandroid.model.Multistatus;
+import com.thegrizzlylabs.sardineandroid.model.Prop;
 import com.thegrizzlylabs.sardineandroid.model.Response;
 
-import org.junit.Ignore;
 import org.junit.Test;
 
 import java.io.ByteArrayInputStream;
@@ -789,5 +789,105 @@ public class MultiStatusResponseHandlerTest {
             // protected acls
             assertTrue(davAcl.getAces().get(4).isProtected());
         }
+    }
+
+    @Test
+    public void testCardDAVResponse() throws  IOException {
+        String xml = "<?xml version='1.0' encoding='utf-8'?>\n" +
+                "<multistatus xmlns=\"DAV:\" xmlns:CR=\"urn:ietf:params:xml:ns:carddav\" xmlns:ns2=\"http://inf-it.com/ns/ab/\">\n" +
+                "    <response>\n" +
+                "        <href>/leoandroid01/a/</href>\n" +
+                "        <propstat>\n" +
+                "            <prop>\n" +
+                "                <principal-collection-set>\n" +
+                "                    <href>/</href>\n" +
+                "                </principal-collection-set>\n" +
+                "                <current-user-principal>\n" +
+                "                    <href>/leoandroid01/</href>\n" +
+                "                </current-user-principal>\n" +
+                "                <current-user-privilege-set>\n" +
+                "                    <privilege>\n" +
+                "                        <read />\n" +
+                "                    </privilege>\n" +
+                "                    <privilege>\n" +
+                "                        <all />\n" +
+                "                    </privilege>\n" +
+                "                    <privilege>\n" +
+                "                        <write />\n" +
+                "                    </privilege>\n" +
+                "                    <privilege>\n" +
+                "                        <write-properties />\n" +
+                "                    </privilege>\n" +
+                "                    <privilege>\n" +
+                "                        <write-content />\n" +
+                "                    </privilege>\n" +
+                "                </current-user-privilege-set>\n" +
+                "                <supported-report-set>\n" +
+                "                    <supported-report>\n" +
+                "                        <report>\n" +
+                "                            <expand-property />\n" +
+                "                        </report>\n" +
+                "                    </supported-report>\n" +
+                "                    <supported-report>\n" +
+                "                        <report>\n" +
+                "                            <principal-search-property-set />\n" +
+                "                        </report>\n" +
+                "                    </supported-report>\n" +
+                "                    <supported-report>\n" +
+                "                        <report>\n" +
+                "                            <principal-property-search />\n" +
+                "                        </report>\n" +
+                "                    </supported-report>\n" +
+                "                    <supported-report>\n" +
+                "                        <report>\n" +
+                "                            <sync-collection />\n" +
+                "                        </report>\n" +
+                "                    </supported-report>\n" +
+                "                    <supported-report>\n" +
+                "                        <report>\n" +
+                "                            <CR:addressbook-multiget />\n" +
+                "                        </report>\n" +
+                "                    </supported-report>\n" +
+                "                    <supported-report>\n" +
+                "                        <report>\n" +
+                "                            <CR:addressbook-query />\n" +
+                "                        </report>\n" +
+                "                    </supported-report>\n" +
+                "                </supported-report-set>\n" +
+                "                <resourcetype>\n" +
+                "                    <CR:addressbook />\n" +
+                "                    <collection />\n" +
+                "                </resourcetype>\n" +
+                "                <owner>\n" +
+                "                    <href>/leoandroid01/</href>\n" +
+                "                </owner>\n" +
+                "                <getetag>\"f897be958cef754667b27f56f16eebbc\"</getetag>\n" +
+                "                <getlastmodified>Mon, 03 Sep 2018 17:16:30 GMT</getlastmodified>\n" +
+                "                <getcontenttype>text/vcard</getcontenttype>\n" +
+                "                <getcontentlength>26845</getcontentlength>\n" +
+                "                <displayname>a</displayname>\n" +
+                "                <sync-token>http://radicale.org/ns/sync/57807f4b156e24c14bcd01ed1c853ea8</sync-token>\n" +
+                "                <CR:addressbook-description>vaddressbook-default</CR:addressbook-description>\n" +
+                "                <ns2:addressbook-color>#c5cddaff</ns2:addressbook-color>\n" +
+                "            </prop>\n" +
+                "            <status>HTTP/1.1 200 OK</status>\n" +
+                "        </propstat>\n" +
+                "    </response>\n" +
+                "</multistatus>";
+
+        MultiStatusResponseHandler handler = new MultiStatusResponseHandler();
+        Multistatus status = handler.getMultistatus(new ByteArrayInputStream(xml.getBytes()));
+        assertNotNull(status);
+        assertEquals(1, status.getResponse().size());
+        Response response = status.getResponse().get(0);
+        assertEquals("/leoandroid01/a/", response.getHref());
+        Prop prop = response.getPropstat().get(0).getProp();
+        assertEquals("/", prop.getPrincipalCollectionSet().getHref());
+        assertEquals("/leoandroid01/", prop.getPrincipalURL().getHref());
+        assertEquals(5, prop.getCurrentUserPrivilegeSet().privileges.size());
+        assertNotNull(prop.getResourcetype().getCollection());
+        assertEquals(1, prop.getResourcetype().getAny().size());
+        assertEquals("/leoandroid01/", prop.getOwner().getHref());
+        assertEquals(2, prop.getAny().size());
     }
 }

